Src/library/net/Clib/network.c

```diff
@@ -824,81 +824,188 @@ void c_put_stream(EIF_INTEGER fd, EIF_POINTER stream_pointer, EIF_INTEGER length
 }
 
-EIF_CHARACTER c_read_char(EIF_INTEGER fd, EIF_INTEGER *l_bytes)
-	/*x read a character from socket fd */
+EIF_DOUBLE c_read_double(EIF_INTEGER fd, EIF_INTEGER *l_bytes)
+	/*x read a double from socket fd */
+{
+	EIF_DOUBLE d;
+	d = c_read_double_noexception(fd, l_bytes);
+	eif_net_check(*l_bytes);
+	return d;
+}
+
+EIF_CHARACTER c_read_char_noexception(EIF_INTEGER fd, EIF_INTEGER *l_bytes)
+	/* read a character from socket fd
+	 *	NO exception is raised, and eventual error is return as result!
+	 */
 {
 	char c=0;
-	CREADX(*l_bytes,fd,&c,sizeof(char), 0)
+	*l_bytes = c_recv_noexception(fd, &c, sizeof(char), 0);
 	return (EIF_CHARACTER) c;
 }
 
-EIF_INTEGER c_read_int(EIF_INTEGER fd, EIF_INTEGER *l_bytes)
-	/*x read an integer from socket fd */
+EIF_CHARACTER c_read_char(EIF_INTEGER fd, EIF_INTEGER *l_bytes)
+	/*x read a character from socket fd */
+{
+	EIF_CHARACTER c;
+	c = c_read_char_noexception(fd, l_bytes);
+	eif_net_check(*l_bytes);
+	return c;
+}
+
+EIF_INTEGER c_read_int_noexception(EIF_INTEGER fd, EIF_INTEGER *l_bytes)
+	/* read an integer from socket fd
+	 *	NO exception is raised, and eventual error is return as result!
+	 */
 {
 	EIF_INTEGER i=0L;
-	CREADX(*l_bytes,fd,&i,sizeof(EIF_INTEGER), 0)
+	*l_bytes = c_recv_noexception(fd, &i, sizeof(EIF_INTEGER), 0);
 	return (EIF_INTEGER) ntohl((unsigned long) i);
 }
 
+EIF_INTEGER c_read_int(EIF_INTEGER fd, EIF_INTEGER *l_bytes)
+	/*x read an integer from socket fd */
+{
+	EIF_INTEGER i;
+	i = c_read_int_noexception(fd, l_bytes);
+	eif_net_check(*l_bytes);
+	return i;
+}
+
+EIF_INTEGER c_read_stream_noexception(EIF_INTEGER fd, EIF_INTEGER len, EIF_POINTER buf)
+	/* read a stream of character from socket fd into buffer buf of length len.
+	 *	NO exception is raised, and eventual error is return as result!
+	 */
+{
+	return (EIF_INTEGER) c_recv_noexception(fd, buf, len, 0);
+}
 
 EIF_INTEGER c_read_stream(EIF_INTEGER fd, EIF_INTEGER len, EIF_POINTER buf)
 	/*x read a stream of character from socket fd into buffer buf
 	    of length len */
 {
-	int nr;
-	CREADX(nr,fd,buf,len, 0);
-	return (EIF_INTEGER) nr;
+	return c_receive(fd, buf, len, (EIF_INTEGER) 0);
 }
 
-
 EIF_INTEGER c_receive(EIF_INTEGER fd, EIF_POINTER buf, EIF_INTEGER len, EIF_INTEGER flags)
 	/*x receive at most len bytes from socket fd into buffer buf
 	    flags can be or'ed from 0, MSG_OOB, MSG_PEEK or MSG_DONTROUTE */
 {
-	int result;
-	CREADX(result,fd,buf,len,flags);
-	return (EIF_INTEGER) result;
+	EIF_INTEGER res;
+	res = c_recv_noexception(fd, buf, len, flags);
+	eif_net_check(res);
+	return res;
 }
 
-EIF_INTEGER c_rcv_from(EIF_INTEGER fd, EIF_POINTER buf, EIF_INTEGER len, EIF_INTEGER flags, EIF_POINTER addr, EIF_POINTER addr_len)
-	/*x like c_receive and sender address is stored into socket address
-	    structure addr and address length into *addr_len */
+EIF_INTEGER c_recvfrom_noexception(EIF_INTEGER fd, EIF_POINTER buf, EIF_INTEGER len, EIF_INTEGER flags, EIF_POINTER addr, EIF_POINTER addr_len)
```

Comment:
Could we introduce a macro c_rcv_from , and a new feature c_recvfrom  instead ?
when interfacing with C , it is much nicer to guess quickly the underlying C call...